features/proposed/FxA-76-metrics-opt-out/README.md

```diff
@@ -0,0 +1,194 @@
+for becoming a relier.
+
+### Effects of the preference
+
+For users with `metrics_optout=true`,
+we will disable non-essential metrics collection by:
+
+* Disabling submission of in-page metrics from the content-server
+  (specifically, data currently submitted to the `post-metrics`
+  and `get-metrics-errors` endpoints).
+* Ensuring they are excluded from
+  (or failing that, are in the control group for)
+  any A/B experiments
+
+The preference does *not* opt the user out if metrics gathered
+as a side-effect of operations essential to the service,
+such as aggregated server log metrics
+or the server-side "activity events" metrics stream.
+
+XXX TODO: does this cost us a round-trip to fetch the preference
+or can we inline it with work we're already doing?
+
+### Exposing the preference
+
+XXX TODO: what sort of UI treatment do we want
+for the initial version of this, if any?
+
+### Respecting choices made in Firefox
+
+When accounts.firefox.com is loaded by Firefox,
+it should perform a handshake with the browser
+to discover whether the user
+has opted out of metrics gathering
+via Firefox Health Report.
+If the browser reports that they have done so,
+then we will do our best to respect that preference
+Specifically:
+
+* If there is no currently-logged-in user,
+  we will act as though "metrics\_optout" were true
+  and disable non-essential metrics gathering.
+
+* If there is a currently-logged-in user,
+  and their "metrics\_optout" prefence is `null`,
+  we will update their profile data
+  to set "metrics\_optout=true",
+  so that they don't have to express their preference again.
+
+* If there is a currently-logged-in user,
+  and their "metrics\_optout" prefence is not `null`,
+  we will ignore the value provided by the browser
+  because they've already made an explicit choice.
+
+* If a user creates an account or logs in
+  via a browser with FHR disabled,
+  we will automatically set "metrics\_optout=true"
+  in their profile data.
+
+XXX TODO: this needs more design work
+informed by previous discussions about "device handshake".
+
+The handshake will go something like this:
+
+* Early in its loading sequence, fxa-content-server
+  ends a WebChannel message to probe the browser
+  for internal state.
+* fxa-content-server listens for
+  a response from the browser,
+  timing out after a reasonable period
+  if no reply is received.
+  The timeout behavoir may be adjusted
+  depending on the context parameter.
+* The browser responds with a JSON object.
+  If the preference `datareporting.healthreport.uploadEnabled` is false
+  then the object will include a key "metrics\_optout" with value `true`.
+  If the preference is unset or true then no "metrics\_optout" key is
+  provided.
+
+We may choose to include additional data in the response,
+such as reporting the logged-in state of the browser,
+but the metrics opt-out preference
+is the only piece of data that must ship
+in the initial version of the handshake.
```

Comment:
This implies this work takes precedence over the other handshake work we have discussed. I'd like to discuss further, without some notion of capabilities we are going to have to keep doing the context version bump tango.